neo4j-plugin/src/main/java/io/cdap/plugin/neo4j/source/Neo4jSource.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.neo4j.source;
+
+import com.google.common.base.Preconditions;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.batch.Input;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.api.plugin.PluginProperties;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.batch.BatchRuntimeContext;
+import io.cdap.cdap.etl.api.batch.BatchSourceContext;
+import io.cdap.cdap.etl.api.validation.InvalidConfigPropertyException;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import io.cdap.plugin.common.ReferenceBatchSource;
+import io.cdap.plugin.common.ReferencePluginConfig;
+import io.cdap.plugin.common.SourceInputFormatProvider;
+import io.cdap.plugin.db.ConnectionConfig;
+import io.cdap.plugin.db.ConnectionConfigAccessor;
+import io.cdap.plugin.db.SchemaReader;
+import io.cdap.plugin.util.DBUtils;
+import io.cdap.plugin.util.DriverCleanup;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.hadoop.mapreduce.lib.db.DBConfiguration;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Objects;
+import java.util.Properties;
+
+/**
+ * Batch source to read from Neo4j.
+ */
+@Plugin(type = "batchsource")
+@Name(Neo4jSource.NAME)
+@Description("")
+public class Neo4jSource extends ReferenceBatchSource<LongWritable, Neo4jRecord, StructuredRecord> {
+  public static final String NAME = "Neo4jSource";
+  private static final Logger LOG = LoggerFactory.getLogger(Neo4jSource.class);
+
+  private static final SchemaTypeAdapter SCHEMA_TYPE_ADAPTER = new SchemaTypeAdapter();
+
+  protected Class<? extends Driver> driverClass;
+
+  private final Neo4jSourceConfig config;
+
+  public Neo4jSource(Neo4jSourceConfig config) {
+    super(new ReferencePluginConfig(config.getReferenceName()));
+    this.config = config;
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    driverClass = context.loadPluginClass(getJDBCPluginId());
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
```

Comment:
what was the reason to override most of the abstract class? Is there a way to make abstraction better to avoid code duplicates?

Reply:
For this was several reason:
1. Problem with initialization neo4j-jdbc-driver (throw exception during setting transaction isolation level)
2. In some parts of existing implementation usied SQL syntaxys that not avalible in Neo4j databease.